client/asset/dcr/dcr.go

```diff
@@ -4909,3 +4933,200 @@ func float64PtrStr(v *float64) string {
 	}
 	return strconv.FormatFloat(*v, 'f', 8, 64)
 }
+
+// ConfirmRedemption returns how many confirmations a redemption has. Normally
+// this is very straitforward. However there are two situations that have come
+// up that this also handles. One is when the wallet can not find the redemption
+// transaction. This is most likely because the fee was set too low and the tx
+// was removed from the mempool. In the case where it is not found, this will
+// send a new fee tx using the provided fee suggestion. The second situation
+// this watches for is a transaction that we can find but has been sitting in
+// the mempool for a long time. This has been observed with the wallet in SPV
+// mode and the transaction inputs having been spent by another transaction. The
+// wallet will not pick up on this so we could tell it to abandon the original
+// transaction and, again, send a new one using the provided feeSuggestion, but
+// only warning for now. This method should not be run for the same redemption
+// concurrently as it need to watch a new redeem transaction before finishing.
+func (dcr *ExchangeWallet) ConfirmRedemption(coinID dex.Bytes, redemption *asset.Redemption, feeSuggestion uint64) (*asset.ConfirmRedemptionStatus, error) {
+	txHash, _, err := decodeCoinID(coinID)
+	if err != nil {
+		return nil, err
+	}
+
+	var secretHash [32]byte
+	copy(secretHash[:], redemption.Spends.SecretHash)
+	dcr.mempoolRedeemsMtx.RLock()
+	mRedeem, have := dcr.mempoolRedeems[secretHash]
+	dcr.mempoolRedeemsMtx.RUnlock()
+
+	var deleteMempoolRedeem bool
+	defer func() {
+		if deleteMempoolRedeem {
+			dcr.mempoolRedeemsMtx.Lock()
+			delete(dcr.mempoolRedeems, secretHash)
+			dcr.mempoolRedeemsMtx.Unlock()
+		}
+	}()
+
+	tx, err := dcr.wallet.GetTransaction(dcr.ctx, txHash)
+	if err != nil && !errors.Is(err, asset.CoinNotFoundError) {
+		return nil, fmt.Errorf("problem searching for redemption transaction %s: %w", txHash, err)
+	}
+	if err == nil {
+		if have && mRedeem.txHash == *txHash {
+			if tx.Confirmations == 0 && time.Now().After(mRedeem.firstSeen.Add(maxRedeemMempoolAge)) {
```

Comment:
In this case, shouldn't we check to see if the contract output is still unspent, and resubmitting if it isn't, instead of just returning 0 confirmations?

Reply:
We just found the spending transaction (the redeem) here.  Do you mean go look for a different mined spender of the contract output as we do in the flow later in this function in the case where `dcr.wallet.GetTransaction` could not find the redeem?

The wallet should periodically rebroadcast the unconfirmed redeem that it initially broadcasted.  Abandoning it and making an entirely new transaction is fairly perilous since the outputs from the initial redeem may be used already.

Reply:
Resubmit the same transaction?

Reply:
> Resubmit the same transaction?

It would surely just say "already have transaction" or similar.  `GetTransaction` just reported it.

Reply:
> it says that if the transaction has been sitting in the mempool for a long time, it means that the inputs have been spent by another transaction

That has to be an insane wallet bug.  Sitting in mempool for a long time does not in general mean the inputs were double spent.

> Which basically means another transaction redeemed the swap somehow.  In this case shouldn't we check to see if the contract output is no longer unspent, and if so, return that the redemption is confirmed?

"no longer unspent" doesn't make sense to me.  It will definitely not be unspent the second the redeem tx enters mempool via the wallet.

Reply:
This is what's in the comment:

```
The second situation
// this watches for is a transaction that we can find but has been sitting in
// the mempool for a long time. This has been observed with the wallet in SPV
// mode and the transaction inputs having been spent by another transaction. 
```

Reply:
I don't believe that comment is correct in general.  What I recall is that Joe hit a pathological case where the core startup match resolution actions fired of a refund or redeem before the wallet was synced, tricking wallet to double spend.  We've since resolved this with synchronized checks

Reply:
> If it's not this strange wallet issue, then the reason the tx has been sitting in the mempool for 2 hours must be that the fee is too low. In which case I think we should resubmit with a higher fee.

What do you mean be "resubmit with a higher fee"?  That sounds like RBF, but a Decred node will not relay a double spend, and the wallet will not publish the transaction either (it only got to that failure mode for the reason I described in my previous comment).

Reply:
Ok to punt on this for now? Maybe whatever I experienced is fixed. Maybe better to go further down this road if we can reliably reproduce something.

Reply:
> Ok to punt on this for now? Maybe whatever I experienced is fixed. Maybe better to go further down this road if we can reliably reproduce something.

Yes, I think so.  I think it was an edge case, but even if we want to consider abandoning, that's just a wallet concept and the rest of the network probably already has the initial tx (unless it was this wallet double spend bug) and a new one that tries to spend the already-spent contract output will be rejected by the network.